### DISCUSSION TITLE
Calculating phase equilibria and diagram for systems with ionic species

### DISCUSSION THREAD

**Discussion by yxj135** [pycalphad/pycalphad#385] [Q&A] [answered]
I am interested in calculating phase equilibria and diagram for systems with ionic species using pycalphad. However, I'm not sure if the calculations are correct, especially when a component has multiple ionic species with different valences. Is it possible to add a simple benchmark example for equilibrium and phase diagram calculations involving charged species in "Examples"?

Thanks a lot!

**Comment by bocklund** [comment]
Hi @yxj135 and welcome!

Do you mean for charged ions in the ionic liquid model or in other models (e.g. solid phases)?

The ionic liquid model should work. If not, please open an issue with an example reproducing a point equilibrium calculation that you believe is incorrect.

Charged species in other models (e.g. solid phases) is known to not be supported yet. @huisun24 is working on finishing support for the charge constraints that pycalphad needs to handle charged species correctly in [this branch](https://github.com/HUISUN24/pycalphad/tree/charge_constrains). If this is what you mean, do you have a particular system/database that you are working on that we can test against?

**Reply by bocklund** [reply]
Update: pull request #386 will add support for charged species.

**Reply by richardotis** [reply]
Update: This is now supported as of pycalphad 0.10.0 (released February 21, 2022). Example here: https://github.com/pycalphad/pycalphad/blob/0b246ad9c40d0106d00b5087ac0928fd77669718/examples/ChargedPhases.ipynb

**Comment by yxj135** [comment]
@bocklund Thanks for your quick response. Right now I'm mostly interested in solid phases such as oxides (e.g., ABO3 perovskites with nonstoichiometry) rather than ionic liquids. For testing purposes, I'm testing even a much simpler model system described by the following arbitrary database:

```
$  database_version ABC_2018 

 Element A   Fcc	0  0  0!
 Element B   Bcc	0  0  0!
 Element VA  VACUUM 0  0  0!

 SPECIES  A+1   A1/+1!
 SPECIES  A+2   A1/+2!
 SPECIES  A+3   A1/+3!
 SPECIES  B-2   B1/-2!

 Phase alpha  %  1 1 !
 Constituent alpha  :A, B:  !
 Parameter G(alpha,A;0)  298 0;   6000  N !
 Parameter G(alpha,B;0)  298 0;   6000  N !

 Phase beta  %  2 1 1 !
 Constituent beta  :A+1, A+2, A+3, VA :B-2 :  !
 Parameter G(beta,A+1:B-2;0)  298   +2.078614962*T;  6000  N !
 Parameter G(beta,A+2:B-2;0)  298   -10.39307481*T;  6000  N !
 Parameter G(beta,A+3:B-2;0)  298   +2.078614962*T;  6000  N !
 Parameter G(beta,VA:B-2;0)  298   +0.8314459848*T;  6000  N !
```

The equilibrium compositions of this simple system can be calculated analytically and can be used as benchmark for your testing. Looking forward to the implementation by @HUISUN24!

Thanks!

------
* Edited to add code formatting to the TDB by @bocklund

**Reply by bocklund** [reply]
Thanks! Toy examples like this are great for testing.

**Comment by richardotis** [comment, marked as answer]
@yxj135 This is now supported as of pycalphad 0.10.0 (released February 21, 2022). Example here: https://github.com/pycalphad/pycalphad/blob/0b246ad9c40d0106d00b5087ac0928fd77669718/examples/ChargedPhases.ipynb
